### PR TITLE
Morpheus changes related to utility consolidation

### LINKED ISSUE #87: [FEA] Create Utility Repo to Store Common Functionality
**Is your feature request related to a problem? Please describe.**
We currently have several repos that all require similar functionality related to CMake, CI, Conda, Docker, and testing. This functionality is duplicated making it hard to maintain.

**Describe the solution you'd like**
Create a new repo containing the above common functionality that can be added as a submodule to all repos that need it. This will allow for versioning the common tools and keeping everything in sync. A partial list of items that could be included is:

https://github.com/nv-morpheus/Morpheus/issues/87#tasklist-block-a0cb8f16-f948-4320-88fb-0aba302f1b36

https://github.com/nv-morpheus/Morpheus/issues/87#tasklist-block-a75e99f1-0d02-4c7c-8b92-0fa689e61450

https://github.com/nv-morpheus/Morpheus/issues/87#tasklist-block-d65be84f-7972-407c-a3b6-7ffa30ce9d7c

### PR DESCRIPTION
**Pull request by drobison00**
Depends on: https://github.com/nv-morpheus/utilities/pull/1/files#diff-9a3c07061febeba4d3e5191e87706e262644abf8514317eeb084e67137e7c807

This PR is branched from the base Morpheus conversion to MRC (https://github.com/nv-morpheus/Morpheus/pull/530). It may make sense to go in as a monolithic patch, or piecemeal.

Shifts to using a shared `utilities` submodule for shared cmake, scripts, utilities, etc.. This shared repo provides a number of new `morpheus_utils_xxx` cmake macros/functions that are now used throughout the configuration process.

Closes #87

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -2,3 +2,7 @@
 	path = external/morpheus-visualizations
 	url = https://github.com/nv-morpheus/morpheus-visualizations.git
 	branch = branch-22.11
+[submodule "morpheus_utils"]
+	path = external/utilities
+	url = git@github.com:nv-morpheus/utilities.git
+	branch = devin-issue-225-87
```
Comment (drobison00) [inline .gitmodules:8]:
```suggestion
	url = https://github.com/nv-morpheus/utilities.git
	branch = branch-23.01
```

Comment (drobison00) [reply .gitmodules:8]: Roll before merge

File: cmake/dependencies.cmake [hunk 2]
```diff
@@ -83,37 +83,30 @@ endif()
 
 # libcudacxx -- get an explicit lubcudacxx build, matx tries to pull a tag that doesn't exist.
 # =========
-set(LIBCUDACXX_VERSION "1.8.0" CACHE STRING "Version of libcudacxx to use")
-include(deps/Configure_libcudacxx)
+morpheus_utils_configure_libcudacxx()
```
Comment (mdemoret-nv) [inline cmake/dependencies.cmake:86]: Cant comment on the right line, but is this still needed? 
```
# Needed force rapids cpm to use our source directory.
set(CPM_SOURCE_CACHE "${CMAKE_SOURCE_DIR}/.cache/cpm")
# Prevent cpm_init from trying to tell us where to put cpm.cmake
include(get_cpm)
# Cant use rapids_cpm_init() for now since the `rapids_cpm_download()` creates a
# new scope when importing CPM. Manually do the other commands and import CPM on
# our own with get_cpm
#include("${rapids-cmake-dir}/cpm/detail/load_preset_versions.cmake")
#rapids_cpm_load_preset_versions()
```
Odd that morpheus requires it but MRC doesnt

File: cmake/environment/init_ccache.cmake [hunk 3]
```diff
@@ -13,11 +13,17 @@
...
-# Set the RPATH to allow finding other libraries in the same folder
-set(VCPKG_LINKER_FLAGS "-Wl,-rpath,'$ORIGIN':'$ORIGIN/../lib'")
+morpheus_utils_check_cache_path(MORPHEUS_CACHE_DIR)
+
+# Configure CCache if requested
+if(MORPHEUS_USE_CCACHE)
+  morpheus_utils_initialize_ccache(MORPHEUS_CACHE_DIR)
+endif()
+
+morpheus_utils_initialize_cpm(MORPHEUS_CACHE_DIR)
```
Comment (mdemoret-nv) [inline cmake/environment/init_ccache.cmake:27]: This doesnt seem like it should be in the cache configuration file

File: cmake/environment/init_compiler.cmake [hunk 4]
```diff
@@ -23,31 +23,31 @@ list(APPEND CMAKE_MESSAGE_CONTEXT "compiler")
...
 # Only set the clang-tidy options for our source code targets
-if (MORPHEUS_USE_CLANG_TIDY)
+if(MORPHEUS_USE_CLANG_TIDY)
   set(CMAKE_C_CLANG_TIDY "clang-tidy")
   set(CMAKE_CXX_CLANG_TIDY "clang-tidy")
   message(STATUS "Enabling clang-tidy for targets in project ${PROJECT_NAME}")
-endif ()
+endif()
 
 list(POP_BACK CMAKE_MESSAGE_CONTEXT)
```
Comment (mdemoret-nv) [inline cmake/environment/init_compiler.cmake:53]: This is also stuff that should be moved to utilities.

File: cmake/environment/init_coverage.cmake [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+#configure_codecov <TARGET_NAME>
+#]=======================================================================]
+function(configure_codecov_target target)
+  if(${MORPHEUS_ENABLE_CODECOV} STREQUAL "ON")
+    message(STATUS "Configuring target <${target}> for code coverage.")
+    append_coverage_compiler_flags_to_target("${target}")
+  endif()
+endfunction()
+
+list(POP_BACK CMAKE_MESSAGE_CONTEXT)
```
Comment (mdemoret-nv) [inline cmake/environment/init_coverage.cmake:66]: How come this isnt in utilities?

File: CMakeLists.txt [hunk 6]
```diff
@@ -33,50 +34,76 @@ option(MORPHEUS_USE_IWYU "Enable running include-what-you-use as part of the bui
...
+# Load morpheus utils and update CMake paths
+set(MORPHEUS_UTILS_RAPIDS_CMAKE_VERSION ${MORPHEUS_RAPIDS_CMAKE_VERSION})
+include(morpheus_utils/load)
+
+# Configure project package manager
+morpheus_utils_initialize_package_manager(
+    MORPHEUS_USE_CONDA
+    BUILD_SHARED_LIBS
+    MORPHEUS_VCPKG_TOOLCHAIN
+    MORPHEUS_VCPKG_DEFAULT_BINARY_CACHE
```
Comment (drobison00) [inline CMakeLists.txt:78]:
```suggestion
```

File: morpheus/CMakeLists.txt [hunk 7]
```diff
@@ -20,13 +20,8 @@ option(MORPHEUS_PYTHON_INPLACE_BUILD
...
 
-set(Python3_FIND_VIRTUALENV "FIRST")
-set(Python3_FIND_STRATEGY "LOCATION")
-find_package(Python3 REQUIRED COMPONENTS Development Interpreter NumPy)
-
-# Check if SKlearn is installed
-include(cmake/utils/sklearn_check.cmake)
-include(cmake/utils/python_module_tools.cmake)
+# TODO(Devin): Not currently using this, just adding to make module and lib creation
+morpheus_utils_create_python_package(morpheus-package)
```
Comment (mdemoret-nv) [inline morpheus/CMakeLists.txt:24]: You need the corresponding call to `morpheus_utils_build_python_package` after all of the calls to `add_subdirectory`. Look at MRC or the MRC quickstart for examples.

File: morpheus/_lib/cmake/libraries/cuda_utils.cmake [hunk 8]
```diff
@@ -18,15 +18,16 @@ find_package(pybind11 REQUIRED)
...
+    ${MORPHEUS_LIB_ROOT}/src/utilities/type_util.cu
 )
 
-
 set_target_properties(
-  cuda_utils_objs
+    cuda_utils_objs
     PROPERTIES
+      CUDA_STANDARD 17
+      CUDA_STANDARD_REQUIRED ON
```
Comment (mdemoret-nv) [inline morpheus/_lib/cmake/libraries/cuda_utils.cmake:30]: Why is this necessary?

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
1. `FindSphinx.cmake` should be moved to utilities with the other find macros
2. `get_cpm.cmake` Is this still needed and used? Seems like it would conflict with the utilities stuff
3. The GTest and benchmarks dependencies should be moved to utilities
4. The `morpheus` package needs to use the `morpheus_utils_build_python_package` function to finalize the package.

File: cmake/environment/init_coverage.cmake (hunk 5 above)
Comment (drobison00) [reply cmake/environment/init_coverage.cmake:66]: Its Morpheus specific; all the common coverage items are in the utilities library, but this more morpheus specific in terms of testing for enabling variables, and specifying exclusion directories.

File: morpheus/_lib/cmake/libraries/cuda_utils.cmake (hunk 8 above)
Comment (drobison00) [reply morpheus/_lib/cmake/libraries/cuda_utils.cmake:30]: Because nvcc doesn't support c++20 for the versions we use. If we don't set this it inherits from CMAKE_CXX_STANDARD, and CMake fails because it doesn't know how to set CUDA20 compiler parameters for nvcc.

File: cmake/dependencies.cmake (hunk 2 above)
Comment (drobison00) [reply cmake/dependencies.cmake:86]: I think so; I remember trying to remove it and having problems. I think it might be libcudacxx that is setting `CPM_SOURCE_CACHE`. But I'll double check.

File: docker/conda/environments/cuda11.5_dev.yml [hunk 9]
```diff
@@ -60,6 +62,7 @@ dependencies:
     - mlflow>1.29,<2
     - mrc=23.01
     - myst-parser==0.17
+    - nbsphinx=0.8.10 # For Sphinx documentation
```
Comment (dagardner-nv) [inline docker/conda/environments/cuda11.5_dev.yml:65]: If we are moving the dependencies needed to build the docs into the main conda env file, we should be sure we add all of them and then rm the `docs/requirements.txt` and `docs/conda_docs.yml` files, and update `ci/scripts/github/docs.sh` script.
```suggestion
    - nbsphinx=0.8.10 # For Sphinx documentation
```

**Review by dagardner-nv — CHANGES_REQUESTED** [review]
Can you make the utilities repo public? I can only see what is being moved & removed in this PR, and can't see what is in the utilities repo.

File: docker/install_docker.sh [hunk 10]
```diff
@@ -1,25 +0,0 @@
-#!/bin/bash
```
Comment (dagardner-nv) [inline docker/install_docker.sh:1]: This script is needed, as anyone using the pre-built containers needs to be able to launch Triton from within the Morpheus container (#552 ). 

The existing docker build doesn't include docker itself, and the documentation currently instructs users on running this script.

Comment (dagardner-nv) [reply docker/install_docker.sh:1]: I see this was moved to the utilities repo, we just need to ensure that the  `docs/source/developer_guide/contributing.md` and `docs/source/getting_started.md` docs that reference this script are updated

Comment (drobison00) [reply docker/install_docker.sh:1]: Updated.

File: docker/conda/environments/cuda11.5_dev.yml (hunk 9 above)
Comment (drobison00) [reply docker/conda/environments/cuda11.5_dev.yml:65]: This was the wrong choice on my part. I'll roll these changes back.

File: .gitmodules [hunk 11]
```diff
@@ -4,5 +4,5 @@
 	branch = branch-22.11
 [submodule "morpheus_utils"]
 	path = external/utilities
-    url = https://github.com/nv-morpheus/utilities.git
+	url = ../utilities.git
```
Comment (mdemoret-nv) [inline .gitmodules:7]: To make this work across forks, this needs to be `../../nv-morpheus/utilities.git`

File: ci/scripts/github/common.sh [hunk 12]
```diff
@@ -60,7 +60,7 @@ export SCCACHE_REGION="${AWS_DEFAULT_REGION}"
 export SCCACHE_IDLE_TIMEOUT=32768
 #export SCCACHE_LOG=debug
 
-export CMAKE_BUILD_ALL_FEATURES="-DCMAKE_MESSAGE_CONTEXT_SHOW=ON -DMORPHEUS_BUILD_BENCHMARKS=ON -DMORPHEUS_BUILD_EXAMPLES=ON -DMORPHEUS_BUILD_TESTS=ON -DMORPHEUS_USE_CONDA=ON -DMORPHEUS_PYTHON_INPLACE_BUILD=OFF -DMORPHEUS_BUILD_PYTHON_STUBS=ON -DMORPHEUS_USE_CCACHE=ON"
+export CMAKE_BUILD_ALL_FEATURES="-DCMAKE_MESSAGE_CONTEXT_SHOW=ON -DMORPHEUS_CUDA_ARCHITECTURES=60;70;75;80 -DMORPHEUS_BUILD_BENCHMARKS=ON -DMORPHEUS_BUILD_EXAMPLES=ON -DMORPHEUS_BUILD_TESTS=ON -DMORPHEUS_USE_CONDA=ON -DMORPHEUS_PYTHON_INPLACE_BUILD=OFF -DMORPHEUS_BUILD_PYTHON_STUBS=ON -DMORPHEUS_USE_CCACHE=ON"
```
Comment (mdemoret-nv) [inline ci/scripts/github/common.sh:63]: How come we arent using `all` here?

File: setup.py [hunk 13]
```diff
@@ -21,9 +21,9 @@
 # Note: this is order dependent
 os.environ["PIP_FIND_LINKS"] = "https://download.pytorch.org/whl/cu113/torch_stable.html"
 
-import versioneer
-from setuptools import find_packages
-from setuptools import setup
+import versioneer  # noqa: E402
+from setuptools import find_packages  # noqa: E402
+from setuptools import setup  # noqa: E402
```
Comment (mdemoret-nv) [inline setup.py:26]: Why are these needed?

**Review by mdemoret-nv — APPROVED** [review]
Approved. Just a couple of comments.

File: ci/scripts/github/common.sh (hunk 12 above)
Comment (drobison00) [reply ci/scripts/github/common.sh:63]: All fails in CI because it tries to include pre-pascal architectures.

File: setup.py (hunk 13 above)
Comment (drobison00) [reply setup.py:26]: This makes it possible for setuptools to find the right versions of pytorch and install them as dependencies, because it needs an alternative PIP repository (download.pytorch.org).

If you try to install Morpheus locally, via pip install -e ., it won't work unless you manually specify PIP_FIND_LINKS on the command line, or we inject it here before setuptools is imported. The #noqa flags are to stop flake from complaining that we have code between imports.

Error: 
```bash
ERROR: Could not find a version that satisfies the requirement torch==1.12.0+cu113 (from morpheus) (from versions: 1.4.0, 1.5.0, 1.5.1, 1.6.0, 1.7.0, 1.7.1, 1.8.0, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1)
ERROR: No matching distribution found for torch==1.12.0+cu113
```
